<!DOCTYPE html>
<html lang="en-us">
<head>
    <meta charset="UTF-8">
    <title>Diversity Analyzer</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
</head>
<body>

<section class="page-header">
      <h1 class="project-name" align=center>Diversity Analyzer</h1>
      <h2 class="project-tagline" align=center>Tool for VJ, CDR, and SHM labeling of adaptive immune repertoires</h2>
      <a href="index.html" class="btn"><b>Home</b></a>
      <a href="https://github.com/Immunotools/immunotools" class="btn"><b>View on GitHub</b></a>
      <a href="https://github.com/Immunotools/immunotools/zipball/master" class="btn"><b>Download .zip</b></a>
      <a href="https://github.com/Immunotools/immunotools/tarball/master" class="btn"><b>Download .tar.gz</b></a>
    </section>

<section class="main-content">
    <ul>
        <li><a href = "#general_info">General info</a></li>
        <li><a href = "#output_details">Output details</a></li>
        <li><a href = "#citation">Citation and feedback</a></li>
    </ul>

    <h2 id="general_info">General info</h2>
    <p align = "justify">
        To launch DiversityAnalyzer, type:</br>

        <code>
            ./diversity_analyzer.py -i INPUT_FILE -o OUTPUT_DIRECTORY -l LOCUS
        </code>
        </br>

        <ul>
            <li><b>INPUT_FILE:</b> input sequences of adaptive immune receptors in FASTA/FASTQ format;</li>
            <li><b>OUTPUT_DIR:</b> output directory. 
                Please note that DiversityAnalyzer checks the existence of OUTPUT_DIR, 
                removes it if it exists, and creates an empty directory with name OUTPUT_DIR. 
                So, we recommend not to specify an existing directory as OUTPUT_DIR!</li>
            <li><b>LOCUS:</b> name of IG / TR locus that will be used for alignment. 
                Possible values: ALL / IG / IGH / IGK / IGL / TR / TRA / TRB / TRG / TRD.
        </ul>
    </p>

    <h3>Optional parameters</h3>
    <p align = "justify">
        <table border="0">
            <col width="30%">
            <col width="20%">
            <col width="50%">
        <tr>
            <td><b>Description</b></td>
            <td><b>Option</b></td>
            <td><b>Values</b></td>
        </tr>
        <tr>
            <td>Organism</td>
            <td><code>--org ORGANISM</code></td>
            <td>human / mouse / rat / rabbit / rhesus-monkey / cattle. Default value is human.</td>
        </tr>
        <tr>
            <td>Number of parallel threads</td>
            <td><code>-t / --threads INT</code></td>
            <td>Default value is 16.</td>
        </tr>
        <tr>
            <td>Skip visualization</td>
            <td><code>--skip-plots</code></td>
            <td>By default, DiversityVisualizer visualizes statistics of input sequences. 
                Note that this step might be time-consuming for large input libraries.</td>
        </tr>
        </table>
    </p>

    <h3>Output</h3>
    <p align = "justify">
        Diversity Analyzer performs <a href = "#vj_labeling_output">V / J</a>,
        <a href = "#cdr_labeling_output">CDR</a>,
        and <a href = "#shm_labeling_output">SHM</a> labelings of input sequences;
        computes their <a href = "#diversity_stats">diversity characteristics</a>;
        computes <a href = "#substitution_output">nucleotide and amino acid substitutions</a>;
        visualizes computed statistics and reports <a href = "#html_report">summary in user-friendly HTML format</a>.
    </p>

    <h2 id="output_details">Output details</h2>
    <ol>
        <li><a href = "#output_summary">Summary report</a></li>
        <li><a href = "#vj_labeling_output">VJ labeling output</a></li>
        <li><a href = "#cdr_labeling_output">CDR labeling output</a></li>
        <li><a href = "#shm_labeling_output">SHM labeling output</a></li>
        <li><a href = "#diversity_stats">Diversity statistics</a></li>
        <li><a href = "#substitution_output">Nucleotide and amino acid substitutions</a></li>
        <li><a href = "#html_report">Summary annotation report</a></li>
    </ol>

    <h3 id = "output_summary">1. Summary report</h3>
    <p align = "justify">
        Diversity Analyzer compiles a report in TSV format including basic diversity characteristics of input sequences.
        For each input sequence, report contains information about:
        <ul>
            <li><b>Chain type:</b> IGH / IGK / IGL / TRA / TRB / TRG / TRD.</li>
            <li><b>V and J hits</b>. Gene names are consistent with IMGT database.</li>
            <li><b>Amino acid sequence.</b></li>
            <li><b>Information about productiveness:</b> presence of stop codon (1 or 0); in-frame (1 or 0); productive (1 or 0).</li>
            <li><b>Positions of CDR1, CDR2, and CDR3.</b> Note that all positions are inclusive.</li>
        </ul>
        Example of summary diversity output can be found <a href = "divan_docs/cdr_details.txt">here</a>.
    </p>

    <hr>
    <!------------------------------------------------------------------------>
    <h3 id = "vj_labeling_output">2. VJ labeling output</h3>
    <p align = "justify">
        Diversity Analyzer launches VJ Finder (fast and accurate aligner of Rep-seq reads) for finding V and J hits of input sequences.
        Based on computed VJ labeling, DiversityAnalyzer reports heatmap showing VJ usage:
    </p>

    <p align="center">
        <a href = "divan_docs/vj_heatmap.png">
            <img src = divan_docs/vj_heatmap.png width  = 30%><br>
            Click for preview
        </a>
    </p>
 
    <p align = "justify">
        DiversityAnalyzer also computes usages of V and J genes in input receptor sequences. 
        Note that allelic variations are collapsed for usage computation.
        E.g., both IGHV1-18*01 and IGHV1-18*02 will contribute to usage of IGHV1-18.
    </p>    

    <p align="center">
        <table border="0">
            <col width="50%">
            <col width="50%">
            <tr>
                <td align = center>
                    <a href = "divan_docs/v_usage.png">
                    <img src = divan_docs/v_usage.png width = 100%><br>
                    Click for preview</a>
                </td>
                <td align = center>
                    <a href = "divan_docs/j_usage.png">
                    <img src = divan_docs/j_usage.png width = 100%><br>
                    Click for preview</a>
                </td>
            </tr>
        </table>
    </p>

    <p align = center>
        DiversityAnalyzer also reports computed V and J usages in TXT files:<br>

        <table border = 0>
            <col width="50%">
            <col width="50%">
            <tr>
                <td align = justify>
                    <code>
                    Gene_name       Multiplicity    Frequency<br>
                    IGHV3-11        1               0.00319<br>
                    IGHV3-20        300             0.95847<br>
                    IGHV3-21        5               0.01597<br>
                    IGHV3-23        7               0.02236<br>
                    </code>
                </td>
                <td>
                    <ul>
                        <li><b>Gene:</b> the name of germline V or J gene</li> 
                        <li><b>Multiplicity:</b> the number of input sequences aligned to the germline gene;</li>
                        <li><b>Frequency:</b> the fraction of input sequences aligned to the germline gene.</li>
                    </ul>
                </td>
            </tr>
        </table>
    </p>

    <h4 id = "v_alignments">2.1. V alignments</h4>
    <p align = "justify">
        Diversity Analyzer reports all computed V alignments in FASTA file.
        Each alignment is presented as a pair of gapped sequences of equal length.
        Code below shows an example of V alignments in FASTA format for a single read-gene pair:
    </p>

    <code>
        >INDEX:1|READ:1_merged_read|START_POS:0|END_POS:295
        GAGGTGCAGTTGGTAGAGTCTGGGGGAGGATTGGTCCAGCCGGGGGGGTCCCTGAGACTCTCCTGTGCAGCCTCTGGATTCACCTTTAGTAGTTATTGGATGAGCTGGGTCC
        GCCAGGCTCCAGGGAAGGGGCTGGAGTGGGTGGCCAACATAAAGCAAGATGAAAGTGAGAAATATTATGTGGACTCTGTGAAGGGCCGATTCACCATCTCCAGAGACAACGC
        CAAGAAGTCACTGTCTCTGCGAACGAACAGGCTGAGAGCCGAGTACATGGCTGTGTATTACTGTGCGAGTAC<br>

        >INDEX:1|GENE:IGHV3-7*01|START_POS:0|END_POS:295|CHAIN_TYPE:IGH
        GAGGTGCAGCTGGTGGAGTCTGGGGGAGGCTTGGTCCAGCCTGGGGGGTCCCTGAGACTCTCCTGTGCAGCCTCTGGATTCACCTTTAGTAGCTATTGGATGAGCTGGGTCC
        GCCAGGCTCCAGGGAAGGGGCTGGAGTGGGTGGCCAACATAAAGCAAGATGGAAGTGAGAAATACTATGTGGACTCTGTGAAGGGCCGATTCACCATCTCCAGAGACAACGC
        CAAGAACTCACTGTATCTGCAAATGAACAGCCTGAGAGCCGAGGACACGGCTGTGTATTACTGTGCGAGAGA
    </code>

    <hr>
    <!------------------------------------------------------------------------>
    <h3 id = "cdr_labeling_output">3. CDR labeling output</h3>
    <p align = "justify">
        Diversity Analyzer aligns input sequences against V and J segments from the IMGT database with known positions of CDRs.
        Using alignment, positions of germline CDRs are converted into positions of receptor sequences:
        
        <img src="divan_docs/cdr_computation.png" alt = "cdr_computation">

        <a href = "http://www.imgt.org/IMGTindex/CDR.php">IMGT</a> and
        <a href = "http://www.bioinf.org.uk/abs/#kabatnum">Kabat</a> notations are available for labeling.
        Diversity Analyzer also reports computed CDR1s, CDR2s, and CDR3s in FASTA format.
    </p>

    <h4 id = "cdr_stats_output">3.1. CDR statistics</h4>
    <p align = "justify">
        If a chain is presented by at least 100 input sequences, Diversity Analyzer computes characteristics of its CDRs:
    </p>

    <p align="center">
    <table border="0">
        <col width="50%">
        <col width="50%">

    <tr>
        <td align = justify>
            <div style="width: 380px">
                Upper right: distribution of CDR3 lengths.
                Diversity Analyzer also selects length of CDR3s that presented by the largest number of CDR3s.
                For CDR3s of this length, nucleotide (lower left) and amino acid (lower right) content are visualized.
                Amino acids are colored according to their hydrophobicity.
                Hydrophobic amino acids are colored in red, hydrophilic amino acids are colored in blue,
                neutral amino acids are colored in pale colors.
            </div>
        </td>

        <td align = center>
            <div style="width: 380px">
                <a href = "divan_docs/IGH_CDR3_length.png">
                    <img src = "divan_docs/IGH_CDR3_length.png" alt = "cdr_nucl_length">
                    Click for preview
                </a>
            </div>
        </td>
    </tr>

        <tr>
            <td align = center>
                <div style="width: 380px">
                    <a href = "divan_docs/IGH_CDR3_nucls.png">
                        <img src = "divan_docs/IGH_CDR3_nucls.png" alt = "cdr_nucl_content">
                        Click for preview
                    </a>
                </div>
            </td>
            <td align = center>
                <div style="width: 380px">
                    <a href = "divan_docs/IGH_CDR3_aa.png">
                        <img src = "divan_docs/IGH_CDR3_aa.png" alt = "aa_nucl_content">
                        Click for preview
                    </a>
                </div>
        </tr>
    </table>
    </p>

    <hr>
    <!------------------------------------------------------------------------>
    <h3 id = "shm_labeling_output">4. SHM labeling output</h3>
    <p align = "justify">
        For each input sequence, Diversity Analyzer computes SHMs and compiles their characteristics in TSV format.
        For each SHM, the following characteristics are reported:
        <ul>
            <li><b>SHM type:</b> substitution (S) / insertion (I) / deletion (D).</li>
            <li><b>Positions on sequence and germline gene.</b></li>
            <li><b>Nucleotides of sequence and germline gene at the position of SHM.</b>
                In case of indel, SHM is denoted as '–'.</li>
            <li><b>Amino acids of sequence and germline gene at the position of SHM.</b></li>
            <li><b>Indicator of SHM synonymy:</b> 1 or 0.</li>
            <li><b>Indicator whether SHM leads to occurrence of stop codon</b>: 1 or 0.</li>
        </ul>

        Lines corresponding to SHMs from different sequences are separated by lines containing information
        about names of input sequence and germline gene, types of segment and chain.
        Example of SHM report for a single sequence is given below:
    </p>

    <p align = center>
        <a href = "divan_docs/shm_details_example.jpg">
            <img src="divan_docs/shm_details_example.jpg" width = 60%><br>
            Click for preview
        </a>
    </p>

    <h4>4.1 SHM plots</h4>
    <p align = justify>
        If a chain presented in input sequences,
        Diversity Analyzer reports plots for SHMs lying in V segments of sequences of this chain:
    </p>

    <table>
        <tr align = center>
            <td><b>IGH</b></td>
            <td><b>IGK</b></td>
            <td><b>IGL</b></td>
        </tr>
        <tr align = center>
            <td>
                <a href = "divan_docs/mutations_distribution_IGHV.png">
                    <img src = divan_docs/mutations_distribution_IGHV.png>
                    Click for preview
                </a>
            </td>
            <td>
                <a href = "divan_docs/mutations_distribution_IGKV.png">
                    <img src = divan_docs/mutations_distribution_IGKV.png>
                    Click for preview
                </a>
            </td>
            <td>
                <a href = "divan_docs/mutations_distribution_IGLV.png">
                    <img src = divan_docs/mutations_distribution_IGLV.png>
                    Click for preview
                </a>
            </td>
        </tr>
    </table>

    <p>
        Diversity Analyzer reports characteristics of special SHMs (indels, synonymous) lying in V segments:
    </p>
    <table>
        <tr align = center>
            <td>
                <b>Distribution of indel SHMs lengths</b>
            </td>
        </tr>
        <tr align = center>
            <td>
                <a href = "divan_docs/indel_shms_length.png">
                    <img src="divan_docs/indel_shms_length.png" width="70%"><br>
                    Click for preview
                </a>
            </td>
        </tr>
    </table>

    <hr>
    <!------------------------------------------------------------------------>

    <h3 id = "diversity_stats">5. Diversity statistics</h3>

    <p align = "justify">
        For each computed CDR3, Diversity Analyzer computes its <i>abundance</i> as a number of input sequences with this CDR3.
        For a collection of CDR3 abundances, Diversity Analyzer applies methods of population analysis for computation of diversity characteristics of input sequences.
    </p>

    <table align = center border="0">
        <col width="50%">
        <col width="50%">

        <tr align = center>
            <td><b>Simpson index (SI)</b></td>
            <td><b>Clonal Simpson index (CSI)</b></td>
        </tr>
        <tr>
            <td>
                SI is computed as a probability of two randomly selected CDR3s to be equal.
            </td>
            <td>
                CSI is an approximation of a probability of two randomly selected CDR3s to belong to the same clonal lineage.
            </td>
        </tr>
    </table>

    <p align = "justify">
        A ratio of CSI to SI demonstrates relation between clonally related receptor sequences to random sequences.
        The higher value of CSI / SI, the higher clonal diversity of an input dataset.
        Example of the comparative analysis of three datasets with various properties is given below:
    </p>

    <table align="center" border="0">
        <col width=33%>

        <tr align = center>
            <td><b>PBMCs</b></td>
            <td><b>ASCs (negative to influenza <a href = "https://en.wikipedia.org/wiki/Hemagglutinin">HA</a>)</b></td>
            <td><b>ASCs (positive to influenza <a href = "https://en.wikipedia.org/wiki/Hemagglutinin">HA</a>)</b></td>
        </tr>
        <tr align = center>
            <td>0-th day of vaccination</td>
            <td>7-th day of vaccination</td>
            <td>7-th day of vaccination</td>
        </tr>
        <tr align = center>
            <td>SRR3620098</td>
            <td>SRR3620075</td>
            <td>SRR3620074</td>
        </tr>
        <tr align = center>
            <td>Low clonal diversity</td>
            <td>Medium clonal diversity</td>
            <td>High clonal diversity</td>
        </tr>
        <tr align = center>
            <td><font color="red"><b>CSI / SI = 1.36</b></font></td>
            <td><font color = "orange"><b>CSI / SI = 7.15</b></font></td>
            <td><font color = "#32cd32"><b>CSI / SI = 11.96</b></font></td>
        </tr>
    </table>

    <hr>
    <!------------------------------------------------------------------------>

    <h3 id = "substitution_output">6. Amino acid and nucleotide substitutions</h3>

    <p align = "justify">
        Diversity Analyzer reports and visualizes matrices of nucleotide and amino acid substitutions:
    </p>

    <table>
        <col width = 50%>
        <col width = 50%>

        <tr>
            <td align="center">
                <a href = "divan_docs/nucl_substitutions.png">
                    <img src = divan_docs/nucl_substitutions.png>
                    Click for preview
                </a>
            </td>
            <td align="center">
                <a href = "divan_docs/aa_substitutions.png">
                    <img src = divan_docs/aa_substitutions.png>
                    Click for preview
                </a>
            </td>
        </tr>
    </table>

    <hr>
    <!------------------------------------------------------------------------>

    <h3 id = "html_report">7. Summary annotation report</h3>
    <p align = "justify">
        Diversity Analyzer compiles all computed statistics and plots into a single report in HTML format.
        Example of such report can be found <a href = "divan_docs/annotation_report.html">here</a>
        (antibody repertoire from <a href = "https://zenodo.org/record/10863#.WLxbkBKGPfY">HIV-infected patient, 181 week of disease</a>).
    </p>

    <h2 id = "citation">Citation and feedback</h2>
    <p align = "justify">
        If you use DiversityAnalyzer in your research, please cite <a href = http://www.jimmunol.org/content/early/2017/10/04/jimmunol.1700485>our paper</a>:
        Shlemov A, Bankevich S, Bzikadze A, Turchaninova MA, Safonova Y, Pevzner PA. Reconstructing Antibody Repertoires from Error-Prone Immunosequencing Reads. J Immunol. 2017 Nov 1;199(9):3369-3380.
    </p>

    <p align = "justify">
       If you have any questions or troubles with running DiversityAnalyzer, please contact <a href = "mailto:safonova.yana@gmail.com">Yana Safonova</a>. 
        We also will be happy to hear your suggestions about improvement of our tools!
    </p>

    <hr>
    <!------------------------------------------------------------------------>
</section>
